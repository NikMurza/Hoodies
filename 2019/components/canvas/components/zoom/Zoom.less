@import (reference) '~@common/styles/constants/index';

.visual-editor {
  &__zoom {
    position: fixed;
    bottom: 32px;
    right: 32px;
    z-index: 100;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;

    &__controls {
      display: flex;
      flex-direction: column;
    }

    &-button {
      margin: 0 0 1px;
      background: #ffffff;
      border-radius: 2px;
      box-shadow: 0 8px 12px rgba(115, 115, 115, 0.12);

      &_active {
        color: @blue;
      }

      &-padding {
        margin: 0 0 32px;
      }
    }
    &__value-button {
      background-color: @white;
      box-shadow: 0 0 2px rgba(0, 0, 0, 0.25);
    }
    &-value-button {
      width: 57px;
      padding: 0;
      transition: color 0.15s ease-out;
      border: none;

      &_alt {
        color: @dark-orange;
      }
      &_shift {
        color: @ultramarine;
      }
      &_ctrl {
        color: @green;
      }
    }
  }
}
