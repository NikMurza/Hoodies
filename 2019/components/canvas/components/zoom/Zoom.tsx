import * as React from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';
import classNames from 'classnames';
import * as defaults from '@common/components/lib/canvas/visual-editor/defaults';
import Button from '@common/components/lib/controls/button/Button';
import SVGIcon from '@common/components/lib/controls/svg-icon/SVGIcon';
import Hint from '@common/components/lib/controls/hint/Hint';
import { VisualEditorPosition } from '@common/components/lib/types/visual-editor';
import { getActionIconPath } from '@common/api/utils/object';
import './Zoom.less';

export enum DisplayType {
  Controls,
  ValueButton,
}

export interface ZoomProps extends WithTranslation {
  position: VisualEditorPosition;
  zoom: number;
  onChange(zoom: number): void;
  /** Целевой контейнер для зума */
  container?: HTMLElement;
  min?: number;
  max?: number;
  step?: number;
  displayType?: DisplayType;
  className?: string;
  /**
   * Зумирование происходит только если курсор на свободной области канвы
   */
  zoomOnFreeAreaOnly?: boolean;
}

const cls = 'visual-editor__zoom';

export const Zoom: React.FC<ZoomProps> = (props: ZoomProps): JSX.Element => {
  const {
    position,
    zoom,
    onChange,
    displayType,
    className,
    t,
    container,
    min,
    max,
    step,
    zoomOnFreeAreaOnly,
  } = props;

  const [isCtrlPressed, setIsCtrlPressed] = React.useState(false);
  const [isShiftPressed, setIsShiftPressed] = React.useState(false);
  const [isAltPressed, setIsAltPressed] = React.useState(false);

  const [isModifierPressed, setIsModifierPressed] = React.useState(false);

  const onKeydown = React.useCallback((e: KeyboardEvent): void => {
    setIsCtrlPressed(e.ctrlKey);
    setIsShiftPressed(e.shiftKey);
    setIsAltPressed(e.altKey);

    setIsModifierPressed(e.ctrlKey || e.shiftKey || e.altKey);

    if (e.key === 'Control' || e.key === 'Shift' || e.key === 'Alt') {
      e.preventDefault();
    }
  }, []);

  const onKeyup = React.useCallback((): void => {
    setIsCtrlPressed(false);
    setIsShiftPressed(false);
    setIsAltPressed(false);

    setIsModifierPressed(false);
  }, []);

  React.useEffect(() => {
    document.addEventListener('keydown', onKeydown);
    return (): void => {
      document.removeEventListener('keydown', onKeydown);
    };
  }, []);

  React.useEffect(() => {
    document.addEventListener('keyup', onKeyup);
    return (): void => {
      document.removeEventListener('keyup', onKeyup);
    };
  }, []);

  const onZoomPlus = (): void => {
    const newValue = Math.round((zoom + step) * 100) / 100;
    onChange(Math.min(newValue, max));
  };

  const onZoomMinus = (): void => {
    const newValue = Math.round((zoom - step) * 100) / 100;
    onChange(Math.max(newValue, min));
  };

  const onWheel = React.useCallback(
    (e: React.WheelEvent<HTMLElement> | WheelEvent): void => {
      const { deltaY, deltaX } = e;
      if (deltaX || isModifierPressed) return;
      if (deltaY >= 0) {
        onZoomMinus();
      } else {
        onZoomPlus();
      }
    },
    [zoom, isModifierPressed],
  );

  const onWheelListener = React.useCallback(
    (e: WheelEvent): void => {
      // разрешаем масштабировать только когда курсор на свободной области канваса
      if (zoomOnFreeAreaOnly && (e.target as HTMLElement).closest('.visual-canvas__components')) {
        e.preventDefault();
        return;
      }

      // preventDefault, запрещающий скролл, не должен вызываться на onWheel кнопки со значением,
      // поэтому вынесен отдельно
      if (!isModifierPressed) {
        e.preventDefault();
      }
      onWheel(e);
    },
    [zoomOnFreeAreaOnly, zoom, isModifierPressed],
  );

  const toDefaultValue = (): void => {
    onChange(1);
  };

  React.useEffect(() => {
    container?.addEventListener('wheel', onWheelListener);
    return (): void => {
      container?.removeEventListener('wheel', onWheelListener);
    };
  }, [container, zoom, isModifierPressed]);

  const x = position?.x ?? 0;
  const y = position?.y ?? 0;
  const style: React.CSSProperties = {
    top: y > 0 ? y : undefined,
    bottom: y < 0 ? -y : undefined,
    left: x > 0 ? x : undefined,
    right: x < 0 ? -x : undefined,
  };

  const valueButtonClass = classNames({
    [`${cls}-value-button`]: true,
    [`${cls}-value-button_alt`]: isAltPressed,
    [`${cls}-value-button_shift`]: isShiftPressed,
    [`${cls}-value-button_ctrl`]: isCtrlPressed,
  });

  const renderZoomComponent = (): JSX.Element => {
    const components = {
      [DisplayType.Controls]: (
        <div className={`${cls}__controls`}>
          <Button onClick={onZoomPlus} icon className={`${cls}-button`} size="S">
            <SVGIcon src={getActionIconPath('plus')} />
          </Button>
          <Button
            onClick={onZoomMinus}
            icon
            className={`${cls}-button ${cls}-button-padding`}
            size="S"
          >
            <SVGIcon src={getActionIconPath('minus')} />
          </Button>
        </div>
      ),
      [DisplayType.ValueButton]: (
        <div className={`${cls}__value-button`} onWheel={onWheel}>
          <Hint
            position="top"
            trigger="hover"
            tooltip={t('BackToDefault')}
            withArrow
            tooltipClassName={classNames('ui-hint_small-arrow', {
              'ui-hint_hidden': zoom === 1,
            })}
            triggerClassName="ui-hint_trigger-full-height"
          >
            <Button outline color="default" className={valueButtonClass} onClick={toDefaultValue}>
              {`${Math.round(zoom * 100)}%`}
            </Button>
          </Hint>
        </div>
      ),
    };
    return components[displayType];
  };

  return (
    <div className={classNames(cls, className)} style={style}>
      {renderZoomComponent()}
    </div>
  );
};

Zoom.defaultProps = {
  min: defaults.canvas.zoom.min,
  max: defaults.canvas.zoom.max,
  step: defaults.canvas.zoom.step,
  displayType: DisplayType.Controls,
};

export default withTranslation('zoom')(Zoom);
