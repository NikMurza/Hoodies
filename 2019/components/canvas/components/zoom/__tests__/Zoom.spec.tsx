import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Zoom, ZoomProps, DisplayType } from '@common/components/lib/canvas/components/zoom/Zoom';

describe('Zoom component', () => {
  it('renders correctly - type Controls', () => {
    const x = 10;
    const y = 20;
    const onChange = jest.fn();

    const props: ZoomProps = {
      zoom: 1,
      position: {
        x,
        y,
      },
      onChange,
      t: jest.fn(),
      i18n: null,
      tReady: true,
    };

    const onChangeSpy = jest.spyOn(props, 'onChange');

    const wrapper = shallow(<Zoom {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();

    const buttons = wrapper.find('Button');
    buttons.get(0).props.onClick();
    expect(onChangeSpy).toBeCalledWith(1.1);

    onChangeSpy.mockClear();

    buttons.get(1).props.onClick();
    expect(onChangeSpy).toBeCalledWith(0.9);
  });

  it('renders correctly - type ValueButton', () => {
    const x = 10;
    const y = 20;
    const onChange = jest.fn();

    const props: ZoomProps = {
      zoom: 1,
      position: {
        x,
        y,
      },
      onChange,
      t: jest.fn(),
      i18n: null,
      tReady: true,
      displayType: DisplayType.ValueButton,
      min: 0.25,
      max: 2,
      step: 0.03,
      container: document.querySelector('div'),
    };

    const wrapper = shallow(<Zoom {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();

    const valueButtonWrapper = wrapper.find('.visual-editor__zoom__value-button');
    valueButtonWrapper.simulate('wheel', { deltaY: 1 });
    expect(props.onChange).toHaveBeenCalledWith(0.97);

    valueButtonWrapper.simulate('wheel', { deltaY: -1 });
    expect(props.onChange).toHaveBeenCalledWith(1.03);

    const button = wrapper.find('Button');
    button.simulate('click');
    expect(props.onChange).toHaveBeenCalledWith(1);
  });
});
