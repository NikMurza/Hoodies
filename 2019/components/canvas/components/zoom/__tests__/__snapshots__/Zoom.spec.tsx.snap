// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zoom component renders correctly - type Controls 1`] = `
<div
  className="visual-editor__zoom"
  style={
    Object {
      "bottom": undefined,
      "left": 10,
      "right": undefined,
      "top": 20,
    }
  }
>
  <div
    className="visual-editor__zoom__controls"
  >
    <Button
      className="visual-editor__zoom-button"
      color="default"
      icon={true}
      onClick={[Function]}
      outline={false}
      size="S"
      title=""
      type="button"
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </Button>
    <Button
      className="visual-editor__zoom-button visual-editor__zoom-button-padding"
      color="default"
      icon={true}
      onClick={[Function]}
      outline={false}
      size="S"
      title=""
      type="button"
    >
      <SVGIcon
        loaderSize={24}
        src="test-file-stub"
      />
    </Button>
  </div>
</div>
`;

exports[`Zoom component renders correctly - type ValueButton 1`] = `
<div
  className="visual-editor__zoom"
  style={
    Object {
      "bottom": undefined,
      "left": 10,
      "right": undefined,
      "top": 20,
    }
  }
>
  <div
    className="visual-editor__zoom__value-button"
    onWheel={[Function]}
  >
    <Hint
      color="black"
      disabled={false}
      mode="hint"
      position="top"
      tooltipClassName="ui-hint_small-arrow ui-hint_hidden"
      trigger="hover"
      triggerClassName="ui-hint_trigger-full-height"
      withArrow={true}
    >
      <Button
        className="visual-editor__zoom-value-button"
        color="default"
        icon={false}
        onClick={[Function]}
        outline={true}
        size="M"
        title=""
        type="button"
      >
        100%
      </Button>
    </Hint>
  </div>
</div>
`;
